src/processor/reconnection-graph.service.ts

```diff
@@ -94,41 +93,44 @@ export class ReconnectionGraphService {
       const exportedUpdates = graphState.exportUserGraphUpdates(dsnpUserId.toString());
 
       const providerKeys = createKeys(this.configService.getProviderAccountSeedPhrase());
-      const blockDelay = BlockchainConstants.SECONDS_PER_BLOCK * MILLISECONDS_PER_SECOND;
-      let statefulStorageTxHash: Hash;
-      let txMonitorJob: ITxMonitorJob;
-      let tx: SubmittableExtrinsic<'rxjs', ISubmittableResult>;
+      let batch: SubmittableExtrinsic<'rxjs', ISubmittableResult>[] = [];
       // eslint-disable-next-line no-restricted-syntax
       for (const bundle of exportedUpdates) {
         switch (bundle.type) {
           case 'PersistPage':
-            tx = this.blockchainService.createExtrinsicCall(
-              { pallet: 'statefulStorage', extrinsic: 'upsertPage' },
-              bundle.ownerDsnpUserId,
-              bundle.schemaId,
-              bundle.pageId,
-              bundle.prevHash,
-              Array.from(Array.prototype.slice.call(bundle.payload)),
+            batch.push(
+              this.blockchainService.createExtrinsicCall(
+                { pallet: 'statefulStorage', extrinsic: 'upsertPage' },
+                bundle.ownerDsnpUserId,
+                bundle.schemaId,
+                bundle.pageId,
+                bundle.prevHash,
+                Array.from(Array.prototype.slice.call(bundle.payload)),
+              ),
             );
-            // eslint-disable-next-line no-await-in-loop
-            statefulStorageTxHash = await this.processSingleBatch(providerKeys, tx);
-            txMonitorJob = {
-              id: jobId,
-              txHash: statefulStorageTxHash,
-              lastFinalizedBlockHash,
-            };
-            this.logger.debug(`Adding job to graph change notify queue: ${txMonitorJob.id}`);
-            this.graphTxMonitorQueue.add(`Graph Change Notify Job - ${txMonitorJob.id}`, txMonitorJob, {
-              delay: blockDelay,
-            });
+
+            if (batch.length === this.capacityBatchLimit) {
+              // eslint-disable-next-line no-await-in-loop
+              await this.processSingleBatch(jobId, providerKeys, batch);
+              doTrack = true;
+              batch = [];
+            }
             break;
           default:
             break;
         }
       }
-      return {};
+
+      if (batch.length > 0) {
+        await this.processSingleBatch(jobId, providerKeys, batch);
+        doTrack = true;
+      }
+
+      return doTrack;
     } catch (err) {
       this.logger.error(`Error updating graph for user ${dsnpUserStr}, provider ${providerStr}: ${(err as Error).stack}`);
+      // In case any transaction were submitted, we don't need to track them because we're already going to fail/retry this job
+      await this.cacheService.removeJob(jobId);
```

Comment:
I don't quite get this logic. If transactions were submitted, but are still pending when the job comes around again, wouldn't that cause an issue with the stateful storage nonce?

Reply:
It could, but the app is generally self-healing in these rare scenarios. This scenario would be one where  *some* transactions were submitted but we also threw an error at some point (could only happen if multiple batches are submitted and at least the first batch is submitted ok, but a subsequent batch has an immediate _synchronous_ RPC error or network failure)

(This is highly unlikely because only rare use graphs will require more than a single capacity batch-worth of transactions)

In this case, we would have to track the successfully-submitted transactions to completion, but then at the end still fail the job for retry. We can track this in the backlog, but I don't think it's much of a priority.

Reply:
☝🏻 note that this would be a very rare scenario, as a user would need to be following > 7k in order for us to submit more than a single capacity batch transaction for their graph.

(though, once we get a provider that supports private friendship, that number will be much lower, since we can fit far fewer PRIds in a graph page). Still, it's unlikely that a user will require > 10 graph pages.

Reply:
Yeah likelihood of that happening is low but worth addressing